### PR TITLE
Document gen

### PR DESCRIPTION
**Pull request by mellenburg**

### CODE REVIEW

File: gen/internals.py [hunk 1]
```diff
@@ -165,10 +172,16 @@ def __repr__(self):
...
+    """ Abstraction for specifying the final set of arguments and how
+    they conditions in which they are required
+    """
     # TODO(cmaloney): Make a better API for working with and managing sub scopes. The current
     # dictionary of dictionaries is really hard to use right.
     def __init__(self, variables=None, sub_scopes=None):
+        """
+        Variables: set of parameters
+        subscopes: mapping of variables to scopes
+        """
```
Comment (branden) [inline gen/internals.py]: Type annotations would be a better way to document these details. https://docs.python.org/3/library/typing.html

File: gen/README.md [hunk 2]
```diff
@@ -1,10 +1,18 @@
 # Generates setup / config files for DC/OS
 
-Generates the files which need to be put on individual hosts in order to install DC/OS with the given configuration, with platform / combo specific quirks.
+## Overview
 
-## Debugging a cluster / watching it come up
-```
-ssh into the desired node
-Watch using the journal `journalctl -f`
+`gen` is a library built to generate the files which need to be put on individual hosts in order to install DC/OS with the given configuration, with platform / combo specific quirks. This is achieved through a custom templating engine whose primary feature is the ability to return the full set of necessary arguments and conditions.
```
Comment (branden) [inline gen/README.md]: Parts of this paragraph are unclear to me. How about something like:

> `gen` is the library responsible for generating static configuration for DC/OS cluster nodes from configuration parameters. This is achieved in part with a custom template engine whose primary feature is generating a configuration parameter schema from a configuration template.

File: gen/internals.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import logging
 from contextlib import contextmanager
 from functools import partial, partialmethod
-from typing import Callable, List, Tuple, Union
+from typing import Any, Callable, List, Dict, Set, Tuple, Union
```
Comment (adam-mesos) [inline gen/internals.py]: In alpha order, `Dict` comes before `List`

File: gen/internals.py [hunk 4]
```diff
@@ -75,8 +75,11 @@ def value_id(value: Union[str, Callable, Late]) -> str:
 
 
 class Setter:
+    """ Class utilized by Source so that user can provide either a string or a
+    function whose arguments will be interpretted as required parameters which
```
Comment (adam-mesos) [inline gen/internals.py]: sp: interpreted

File: gen/internals.py [hunk 5]
```diff
@@ -165,12 +175,18 @@ def __repr__(self):
 
 
 class Target:
-
+    """ Abstraction for specifying the final set of arguments and how
+    they conditions in which they are required
```
Comment (adam-mesos) [inline gen/internals.py]: s/how they conditions/the conditions/

File: gen/internals.py [hunk 6]
```diff
@@ -165,12 +175,18 @@ def __repr__(self):
...
+    """
     # TODO(cmaloney): Make a better API for working with and managing sub scopes. The current
     # dictionary of dictionaries is really hard to use right.
-    def __init__(self, variables=None, sub_scopes=None):
-        self.variables = variables if variables else set()
-        self.sub_scopes = sub_scopes if sub_scopes else dict()
+    def __init__(self, variables: Set[str]=set(), sub_scopes: Dict[str, Scope]=dict()):
+        """
+        Variables: set of parameters that must be extracted from Source
+        subscopes: mapping of variables to be conditionally added
```
Comment (adam-mesos) [inline gen/internals.py]: s/Variables/variables/?
s/subscopes/sub_scopes/?

File: gen/internals.py [hunk 7]
```diff
@@ -165,10 +175,16 @@ def __repr__(self):
...
+    """ Abstraction for specifying the final set of arguments and how
+    they conditions in which they are required
+    """
     # TODO(cmaloney): Make a better API for working with and managing sub scopes. The current
     # dictionary of dictionaries is really hard to use right.
-    def __init__(self, variables=None, sub_scopes=None):
+    def __init__(self, variables: Set[str]=None, sub_scopes: Dict[str, Scope]=None):
+        """
+        Variables: set of parameters that must be extracted from Source
+        subscopes: mapping of variables to be conditionally added
```
Comment (adam-mesos) [inline gen/internals.py]: s/Variables/variables/?
s/subscopes/sub_scopes/?

**Comment by mellenburg** [comment]
@adam-mesos @branden Thanks for the feedback; comments addressed. PTAL

File: gen/internals.py [hunk 8]
```diff
@@ -128,9 +131,16 @@ def make_id(self):
 
 
 class Scope:
-    def __init__(self, name: str, cases=None):
+    """ Abstraction for maintaining the mapping between a parameter and the
+    dependent Targets resulting in each case of parameter value. += operator
+    is implemented to allow easily combining Targets without tedious iterating
```
Comment (orsenthil) [inline gen/internals.py]: iteration.

**Comment by mellenburg** [comment]
Squashed and addressed @orsenthil 's minor changes. :ship:

**Pull request merged**